File: Problem_99_Recover_Binary_Search_Tree/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def recover_tree(root: Optional[TreeNode]) -> None:
    first = None
    second = None
    prev = None

    def inorder(node: Optional[TreeNode]) -> None:
        nonlocal first, second, prev
        if not node:
            return

        inorder(node.left)

        if prev is not None and node.val < prev.val:
            if first is None:
                first = prev
            second = node
        prev = node

        inorder(node.right)

    inorder(root)
    if first and second:
        first.val, second.val = second.val, first.val


if __name__ == "__main__":
    root = TreeNode(1, TreeNode(3, None, TreeNode(2)))
    recover_tree(root)
    print(root.left.val, root.left.right.val)

File: Problem_99_Recover_Binary_Search_Tree/solution.cpp
#include <cstddef>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    void recoverTree(TreeNode* root) {
        TreeNode* first = nullptr;
        TreeNode* second = nullptr;
        TreeNode* prev = nullptr;
        inorder(root, first, second, prev);
        if (first && second) {
            swap(first->val, second->val);
        }
    }

private:
    void inorder(TreeNode* node, TreeNode*& first, TreeNode*& second, TreeNode*& prev) {
        if (node == nullptr) {
            return;
        }

        inorder(node->left, first, second, prev);
        if (prev != nullptr && node->val < prev->val) {
            if (first == nullptr) {
                first = prev;
            }
            second = node;
        }
        prev = node;
        inorder(node->right, first, second, prev);
    }
};
